### ISSUE TITLE
Fails to init dev container with usernamespaces enabled in docker.json

### ISSUE THREAD

**Issue by agowa338** [microsoft/vscode-remote-release#7397]
- VSCode Version: 1.72.2
- Local OS Version: Windows_NT x64 10.0.25227
- Remote OS Version: ArchLinux Server with Docker
- Remote Extension/Connection Type: Containers
- Logs:
```
[Warning] One or more build-args [_DEV_CONTAINERS_FEATURE_CONTENT_SOURCE] were not consumed
Successfully built 5af6d5412c4a
Successfully tagged vsc-git_mgmt-87dd506badc5e87333a167fd1faa5b5e:latest
[391177 ms] Start: Run: docker -v
[391200 ms] Start: Run: docker events --format {{json .}} --filter event=start
[391202 ms] Start: Starting container
[391202 ms] Start: Run: docker run --sig-proxy=false -a STDOUT -a STDERR --mount type=volume,src=dind-var-lib-docker,dst=/var/lib/docker --mount type=volume,src=git_mgmt-main-94c26a44277b3cb7a484dca57777b075,dst=/workspaces --mount type=volume,src=vscode,dst=/vscode -l vsch.local.repository=https://github.com/agowa338/git_mgmt.git/tree/main -l vsch.local.repository.volume=git_mgmt-main-94c26a44277b3cb7a484dca57777b075 -l vsch.local.repository.folder=git_mgmt --init --privileged --entrypoint /bin/sh vsc-git_mgmt-87dd506badc5e87333a167fd1faa5b5e -c echo Container started
docker: Error response from daemon: privileged mode is incompatible with user namespaces.  You must run the container in the host namespace when running privileged mode.
See 'docker run --help'.
[391230 ms] Start: Run: docker ps -q -a --filter label=vsch.local.repository=https://github.com/agowa338/git_mgmt.git/tree/main --filter label=vsch.local.repository.volume=git_mgmt-main-94c26a44277b3cb7a484dca57777b075 --filter label=vsch.local.repository.folder=git_mgmt
[390719 ms] Error: Command failed: docker run --sig-proxy=false -a STDOUT -a STDERR --mount type=volume,src=dind-var-lib-docker,dst=/var/lib/docker --mount type=volume,src=git_mgmt-main-94c26a44277b3cb7a484dca57777b075,dst=/workspaces --mount type=volume,src=vscode,dst=/vscode -l vsch.local.repository=https://github.com/agowa338/git_mgmt.git/tree/main -l vsch.local.repository.volume=git_mgmt-main-94c26a44277b3cb7a484dca57777b075 -l vsch.local.repository.folder=git_mgmt --init --privileged --entrypoint /bin/sh vsc-git_mgmt-87dd506badc5e87333a167fd1faa5b5e -c echo Container started
[390719 ms] trap "exit 0" 15
[390719 ms] /usr/local/share/docker-init.sh
[390719 ms] exec "$@"
[390719 ms] while sleep 1 & wait $!; do :; done -
[390719 ms]     at Use (/root/.vscode-remote-containers/dist/dev-containers-cli-0.255.4/dist/spec-node/devContainersSpecCLI.js:1868:1339)
[390719 ms]     at jO (/root/.vscode-remote-containers/dist/dev-containers-cli-0.255.4/dist/spec-node/devContainersSpecCLI.js:1868:1275)
[390719 ms]     at processTicksAndRejections (internal/process/task_queues.js:95:5)
[390719 ms]     at async Xse (/root/.vscode-remote-containers/dist/dev-containers-cli-0.255.4/dist/spec-node/devContainersSpecCLI.js:1874:2030)
[390719 ms]     at async Uf (/root/.vscode-remote-containers/dist/dev-containers-cli-0.255.4/dist/spec-node/devContainersSpecCLI.js:1874:3193)
[390719 ms]     at async Cae (/root/.vscode-remote-containers/dist/dev-containers-cli-0.255.4/dist/spec-node/devContainersSpecCLI.js:1994:10350)
[390719 ms]     at async yae (/root/.vscode-remote-containers/dist/dev-containers-cli-0.255.4/dist/spec-node/devContainersSpecCLI.js:1994:10104)
[390765 ms] Exit code 1
[390766 ms] Start: Run: docker rm -f 95a9642485bb73ec399310e970a8671dbeeaf8806334d285942911824f79c45f
[390770 ms] Command failed: node /root/.vscode-remote-containers/dist/dev-containers-cli-0.255.4/dist/spec-node/devContainersSpecCLI.js up --workspace-folder /workspaces/git_mgmt --workspace-mount-consistency cached --id-label vsch.local.repository=https://github.com/agowa338/git_mgmt.git/tree/main --id-label vsch.local.repository.volume=git_mgmt-main-94c26a44277b3cb7a484dca57777b075 --id-label vsch.local.repository.folder=git_mgmt --log-level debug --log-format json --config /workspaces/git_mgmt/.devcontainer/devcontainer.json --override-config /tmp/devcontainer-96c3744c-467f-4b0d-a6bb-9743cd3f601e.json --default-user-env-probe loginInteractiveShell --mount type=volume,source=git_mgmt-main-94c26a44277b3cb7a484dca57777b075,target=/workspaces,external=true --mount type=volume,source=vscode,target=/vscode,external=true --skip-post-create --update-remote-user-uid-default off --mount-workspace-git-root true --terminal-columns 89 --terminal-rows 16
[390770 ms] Exit code 1
[391270 ms] Container server terminated (code: 137, signal: null).

```

Steps to Reproduce:

1. set `/etc/docker/daemon.json` to `{"debug": false, "log-driver": "journald", "userns-remap": "default", "exec-opts": ["native.cgroupdriver=systemd"]}`
2. Create `/etc/subuid` and `/etc/subgid` accordingly.
3. Try to open a git repository in a dev container.
4. Set permissions on docker socket to 777 (otherwise vs code only shows a permission error while trying to start as it tries to access the docker socket from within a container which has a different namespace and therefore fails if only root and docker group [user] has access).

Does this issue occur when you try this locally?: No
Does this issue occur when you try this locally and all extensions are disabled?: No

**Comment by chrmarti** [comment]
The dev container appears to require `--privileged` which does not work with user namespaces. Could you try with a simpler dev container?

**Comment by agowa338** [comment]
@chrmarti The best thing would be to find a way to not require `--privileged`, but if you have a peek at the documentation just adding `--userns=host` to the command would work in the mean time as long as the user is allowed to use privileged.
https://docs.docker.com/engine/security/userns-remap/#user-namespace-known-limitations

I think this is some internal container and not the actual devcontainer. As the devcontainer definition does *NOT* contain any `--privileged` as you can see here: https://github.com/agowa338/.git_mgmt

**Comment by chrmarti** [comment]
Could you try adding `"runArgs": ["--userns=host"]` to the devcontainer.json and see if that indeed works as you expect?

The `docker-in-docker` feature you add in the devcontainer.json requires `--privileged`.

**Comment by agowa338** [comment]
Oh, ok that's where it's originating from.
Adding userns=host to runArgs solves the error of starting the container.

But userns still isn't fully working afterward. For some reason, vscode has permission errors within the container (not within a mount as far as I can see):
```
[563418 ms] Start: Starting container
[563418 ms] Start: Run: docker run --sig-proxy=false -a STDOUT -a STDERR --mount type=volume,src=dind-var-lib-docker,dst=/var/lib/docker --mount type=volume,src=git_mgmt-main-94c26a44277b3cb7a484dca57777b075,dst=/workspaces --mount type=volume,src=vscode,dst=/vscode -l vsch.local.repository=https://github.com/agowa338/git_mgmt.git/tree/main -l vsch.local.repository.volume=git_mgmt-main-94c26a44277b3cb7a484dca57777b075 -l vsch.local.repository.folder=git_mgmt --init --userns=host --privileged --entrypoint /bin/sh vsc-git_mgmt-87dd506badc5e87333a167fd1faa5b5e -c echo Container started
Container started
[564695 ms] Start: Run: docker ps -q -a --filter label=vsch.local.repository=https://github.com/agowa338/git_mgmt.git/tree/main --filter label=vsch.local.repository.volume=git_mgmt-main-94c26a44277b3cb7a484dca57777b075 --filter label=vsch.local.repository.folder=git_mgmt
[564725 ms] Start: Run: docker inspect --type container cbb1bf69caf2
[564754 ms] Start: Inspecting container
[564754 ms] Start: Run: docker inspect --type container cbb1bf69caf208f70efdd68438f1c5dbb44171d74c3c1544ed6f847c04c316b2
[564782 ms] Start: Run in container: /bin/sh
[564786 ms] Start: Run in container: uname -m
mount: /sys/kernel/security: must be superuser to use mount.
Could not mount /sys/kernel/security.
AppArmor detection and --privileged mode might break.
mount: /tmp: must be superuser to use mount.
[564903 ms] x86_64
[564903 ms] 
[564903 ms] Start: Run in container: (cat /etc/os-release || cat /usr/lib/os-release) 2>/dev/null
[564905 ms] PRETTY_NAME="Debian GNU/Linux 11 (bullseye)"
NAME="Debian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
VERSION_CODENAME=bullseye
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
[564905 ms] 
[564906 ms] Start: Run in container: cat /etc/passwd
[564909 ms] Start: Run in container: test -f '/var/devcontainer/.patchEtcEnvironmentMarker'
[564910 ms] 
[564910 ms] 
[564910 ms] Exit code 1
[564911 ms] Start: Run in container: /bin/sh
[564912 ms] Start: Run in container: test ! -f '/var/devcontainer/.patchEtcEnvironmentMarker' && set -o noclobber && mkdir -p '/var/devcontainer' && { > '/var/devcontainer/.patchEtcEnvironmentMarker' ; } 2> /dev/null
[565007 ms] 
[565007 ms] 
[565007 ms] Start: Run in container: cat >> /etc/environment <<'etcEnvrionmentEOF'
[565009 ms] 
[565009 ms] 
[565009 ms] Start: Run in container: test -f '/var/devcontainer/.patchEtcProfileMarker'
[565010 ms] 
[565010 ms] 
[565010 ms] Exit code 1
[565011 ms] Start: Run in container: test ! -f '/var/devcontainer/.patchEtcProfileMarker' && set -o noclobber && mkdir -p '/var/devcontainer' && { > '/var/devcontainer/.patchEtcProfileMarker' ; } 2> /dev/null
[565013 ms] 
[565013 ms] 
[565013 ms] Start: Run in container: sed -i -E 's/((^|\s)PATH=)([^\$]*)$/\1${PATH:-\3}/g' /etc/profile || true
[565048 ms] 
[565048 ms] 
[564244 ms] Start: Run: docker inspect --type container cbb1bf69caf208f70efdd68438f1c5dbb44171d74c3c1544ed6f847c04c316b2
[564944 ms] Start: Run in container: node /root/.vscode-remote-containers/dist/dev-containers-cli-0.255.4/dist/spec-node/devContainersSpecCLI.js read-configuration --workspace-folder /workspaces/git_mgmt --id-label vsch.local.repository=https://github.com/agowa338/git_mgmt.git/tree/main --id-label vsch.local.repository.volume=git_mgmt-main-94c26a44277b3cb7a484dca57777b075 --id-label vsch.local.repository.folder=git_mgmt --container-id cbb1bf69caf208f70efdd68438f1c5dbb44171d74c3c1544ed6f847c04c316b2 --log-level debug --log-format json --config /workspaces/git_mgmt/.devcontainer/devcontainer.json --override-config /tmp/devcontainer-289a840b-33b1-40ce-add9-c6e4eb895971.json --mount-workspace-git-root true
[566038 ms] (node:307) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
[566039 ms] (Use `node --trace-deprecation ...` to show where the warning was created)
[566899 ms] @devcontainers/cli 0.20.0. Node.js v14.20.1. linux 6.0.5-arch1-1 x64.
[566898 ms] Start: Run: git rev-parse --show-cdup
[566906 ms] Start: Run: docker inspect --type container cbb1bf69caf208f70efdd68438f1c5dbb44171d74c3c1544ed6f847c04c316b2
[566123 ms] Start: Run in container: node /root/.vscode-remote-containers/dist/dev-containers-cli-0.255.4/dist/spec-node/devContainersSpecCLI.js read-configuration --workspace-folder /workspaces/git_mgmt --id-label vsch.local.repository=https://github.com/agowa338/git_mgmt.git/tree/main --id-label vsch.local.repository.volume=git_mgmt-main-94c26a44277b3cb7a484dca57777b075 --id-label vsch.local.repository.folder=git_mgmt --container-id cbb1bf69caf208f70efdd68438f1c5dbb44171d74c3c1544ed6f847c04c316b2 --log-level debug --log-format json --config /workspaces/git_mgmt/.devcontainer/devcontainer.json --override-config /tmp/devcontainer-289a840b-33b1-40ce-add9-c6e4eb895971.json --include-merged-configuration --mount-workspace-git-root true
[567083 ms] (node:335) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
[567083 ms] (Use `node --trace-deprecation ...` to show where the warning was created)
[567944 ms] @devcontainers/cli 0.20.0. Node.js v14.20.1. linux 6.0.5-arch1-1 x64.
[567944 ms] Start: Run: git rev-parse --show-cdup
[567950 ms] Start: Run: docker inspect --type container cbb1bf69caf208f70efdd68438f1c5dbb44171d74c3c1544ed6f847c04c316b2
[567164 ms] Start: Inspecting container
[567165 ms] Start: Run: docker inspect --type container cbb1bf69caf208f70efdd68438f1c5dbb44171d74c3c1544ed6f847c04c316b2
[567582 ms] Start: Run in container: /bin/sh
[567620 ms] Start: Run in container: uname -m
[568294 ms] x86_64
[568294 ms] 
[568295 ms] Start: Run in container: (cat /etc/os-release || cat /usr/lib/os-release) 2>/dev/null
[568297 ms] PRETTY_NAME="Debian GNU/Linux 11 (bullseye)"
NAME="Debian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
VERSION_CODENAME=bullseye
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
[568297 ms] 
[568298 ms] Start: Run in container: cat /etc/passwd
[568302 ms] Start: Run in container: /bin/sh
[568337 ms] Start: Run in container: stat -c %u /workspaces/git_mgmt
[569074 ms] 63046632
[569075 ms] 
[569075 ms] Start: Run in container: stat -c %u /workspaces/git_mgmt/.devcontainer/devcontainer.json
[569077 ms] 63046632
[569077 ms] 
[569078 ms] Start: Updating configuration state
[569103 ms] Start: Setup shutdown monitor
[569103 ms] Forking shutdown monitor: c:\Users\Klaus\.vscode\extensions\ms-vscode-remote.remote-containers-0.255.4\dist\shutdown\shutdownMonitorProcess \\.\pipe\vscode-remote-containers-6a064b8fac2915bd8e13e392d73596813957a447-sock singleContainer Debug c:\Users\Klaus\AppData\Roaming\Code\logs\20221101T091133\exthost2\ms-vscode-remote.remote-containers 1667295865555
[569115 ms] Start: Run in container: test -d /home/vscode/.vscode-server
[569118 ms] 
[569119 ms] 
[569119 ms] Exit code 1
[569119 ms] Start: Run in container: test -d /home/vscode/.vscode-remote
[569122 ms] 
[569122 ms] 
[569122 ms] Exit code 1
[569122 ms] Start: Run in container: test ! -f '/home/vscode/.vscode-server/data/Machine/.writeMachineSettingsMarker' && set -o noclobber && mkdir -p '/home/vscode/.vscode-server/data/Machine' && { > '/home/vscode/.vscode-server/data/Machine/.writeMachineSettingsMarker' ; } 2> /dev/null
[569125 ms] 
[569126 ms] mkdir: cannot create directory ‘/home/vscode/.vscode-server’: Permission denied
[569126 ms] Exit code 1
[569126 ms] Start: Run in container: cat /home/vscode/.vscode-server/data/Machine/settings.json
[569129 ms] 
[569129 ms] cat: /home/vscode/.vscode-server/data/Machine/settings.json: No such file or directory
[569129 ms] Exit code 1
[569129 ms] Start: Run in container: test -d /home/vscode/.vscode-server/bin/d045a5eda657f4d7b676dedbfa7aab8207f8a075
[569131 ms] 
[569132 ms] 
[569132 ms] Exit code 1
[569132 ms] Start: Run in container: test -d /vscode/vscode-server/bin/linux-x64/d045a5eda657f4d7b676dedbfa7aab8207f8a075
[569134 ms] 
[569135 ms] 
[569135 ms] Exit code 1
[569135 ms] Installing VS Code Server for commit d045a5eda657f4d7b676dedbfa7aab8207f8a075
[569136 ms] Start: Installing VS Code Server
[569136 ms] Start: Run in container: mkdir -p /vscode/vscode-server/bin/linux-x64/d045a5eda657f4d7b676dedbfa7aab8207f8a075_1667296434691
[569139 ms] 
[569140 ms] 
[569270 ms] Start: Run in container: (dd iflag=fullblock bs=8192 count=7140 2>/dev/null; dd iflag=fullblock bs=1714 count=1 2>/dev/null) | tar --no-same-owner -xz -C /vscode/vscode-server/bin/linux-x64/d045a5eda657f4d7b676dedbfa7aab8207f8a075_1667296434691
[577506 ms] 
[577506 ms] 
[577507 ms] Start: Run in container: mv -n /vscode/vscode-server/bin/linux-x64/d045a5eda657f4d7b676dedbfa7aab8207f8a075_1667296434691/* /vscode/vscode-server/bin/linux-x64/d045a5eda657f4d7b676dedbfa7aab8207f8a075 && rmdir /vscode/vscode-server/bin/linux-x64/d045a5eda657f4d7b676dedbfa7aab8207f8a075_1667296434691
[577524 ms] 
[577524 ms] 
[577524 ms] Start: Run in container: mkdir -p '/home/vscode/.vscode-server/bin' && ln -snf '/vscode/vscode-server/bin/linux-x64/d045a5eda657f4d7b676dedbfa7aab8207f8a075' '/home/vscode/.vscode-server/bin/d045a5eda657f4d7b676dedbfa7aab8207f8a075'
[577529 ms] 
[577529 ms] mkdir: cannot create directory ‘/home/vscode/.vscode-server’: Permission denied
[577529 ms] Exit code 1
[577529 ms] Start: Run: docker rm -f 93382b57642ec37c8c6e4b9972e9d567251bf6f639ca58992130e6fffe073528
[577533 ms] Command in container failed: mkdir -p '/home/vscode/.vscode-server/bin' && ln -snf '/vscode/vscode-server/bin/linux-x64/d045a5eda657f4d7b676dedbfa7aab8207f8a075' '/home/vscode/.vscode-server/bin/d045a5eda657f4d7b676dedbfa7aab8207f8a075'
[577534 ms] mkdir: cannot create directory ‘/home/vscode/.vscode-server’: Permission denied
[577534 ms] Exit code 1
[578102 ms] Container server terminated (code: 137, signal: null).

```

**Comment by agowa338** [comment]
Tl;Dr: Adding that to the json worked. The new error is a limitation of docker, where it uses the remapped UID and GIDs for the filesystem even though the namespace should be the same as the hosts...

**Comment by chrmarti** [comment]
Not sure what the correct fix would be here.
